src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/auth/securityconfigurer/CookieAndSessionCsrfTokenRepository.java

```diff
@@ -0,0 +1,153 @@
+/**
+ * The MIT License
+ * Copyright (c) 2018 Estonian Information System Authority (RIA),
+ * Nordic Institute for Interoperability Solutions (NIIS), Population Register Centre (VRK)
+ * Copyright (c) 2015-2017 Estonian Information System Authority (RIA), Population Register Centre (VRK)
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.niis.xroad.restapi.auth.securityconfigurer;
+
+import org.springframework.security.web.csrf.CsrfToken;
+import org.springframework.security.web.csrf.CsrfTokenRepository;
+import org.springframework.security.web.csrf.DefaultCsrfToken;
+import org.springframework.util.ReflectionUtils;
+import org.springframework.util.StringUtils;
+
+import javax.servlet.http.Cookie;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.HttpSession;
+
+import java.lang.reflect.Method;
+import java.util.UUID;
+
+/**
+ * Implements (and copies) functionalities from two csrf repositories:
```

Comment:
It would be nice to avoid copying `HttpSessionCsrfTokenRepository` and `CookieCsrfTokenRepository` and instead create / inject instances of those, and delegate method calls to them appropriately.

I am not sure if that can be achived. There was some problem about `if (!StringUtils.isEmpty(tokenValue)) {` in `saveToken` in some situation X (null tokenValue or something). Both `HttpSessionCsrfTokenRepository.saveToken` and  `CookieCsrfTokenRepository.saveToken` do handle situation X (null tokenValue? or are they not called in situation X?) when used by themselves alone, so maybe it would be possible to implement `CookieAndSessionCsrfTokenRepository.saveToken` so that it calls (or does not call) both existing tokenRepositories with correct parameters in situation X?

Leveraging existing Spring classes would be great since there would be less code for us to maintain, and probably less risk of having updates to HttpSessionCsrfTokenRepository and CookieCsrfTokenRepository which could be out of sync with our copied code.

Reply:
I was able to change the implementation to just a wrapper class for both repositories.

Reply:
Very nice! :+1: :1st_place_medal: